test/object-store/sync/flx_role_change.cpp

```diff
@@ -0,0 +1,857 @@
+                        FAIL("Bootstrap occurred on the second realm, which was not expected");
+                    }
+                    return SyncClientHookAction::NoAction;
+                };
+                auto realm_2 = Realm::get_shared_realm(config_2);
+                set_up_realm(realm_2, params.num_emps);
+
+                test_started = true;
+                // The first rule allows access to all records for user 1
+                AppCreateConfig::ServiceRole user1_role{"user 1 role"};
+                user1_role.apply_when = {{"%%user.id", config_1.sync_config->user->user_id()}};
+                // Add two rules, the first applies to user 1 and the second applies to other users
+                test_rules["roles"] = {transform_service_role(user1_role), transform_service_role(general_role)};
+                update_perms_and_verify(*harness, realm_1, test_rules,
+                                        {BootstrapMode::Any, params.num_emps, params.num_mgrs, params.num_dirs});
+
+                // Realm 2 data should not change (and there shouldn't be any bootstrap messages)
+                verify_records(realm_2, params.num_emps, 0, 0);
+
+                // The first rule will be updated to only have access to employee and managers
+                AppCreateConfig::ServiceRole user1_role_2 = user1_role;
+                user1_role_2.document_filters.read = {{"role", {{"$in", {"employee", "manager"}}}}};
+                user1_role_2.document_filters.write = {{"role", {{"$in", {"employee", "manager"}}}}};
+                // Update the first rule for user 1 and verify the data after the rule is applied
+                test_rules["roles"][0] = {transform_service_role(user1_role_2)};
+                update_perms_and_verify(*harness, realm_1, test_rules,
+                                        {BootstrapMode::Any, params.num_emps, params.num_mgrs, 0});
+
+                // Realm 2 data should not change (and there shouldn't be any bootstrap messages)
+                verify_records(realm_2, params.num_emps, 0, 0);
+            }
+        }
+
+        // Add new sections before this
+        SECTION("Pending changes are lost if not allowed after role change") {
+            std::vector<ObjectId> emp_ids;
+            std::vector<ObjectId> mgr_ids;
+            auto config = harness->make_test_file();
+            config.sync_config->error_handler = [&](std::shared_ptr<SyncSession>, SyncError error) {
+                REQUIRE(!error.is_fatal); // No fatal errors please
+                // Expecting a compensating write error
+                REQUIRE(error.status == ErrorCodes::SyncCompensatingWrite);
+            };
+            auto test_realm = Realm::get_shared_realm(config);
+            set_up_realm(test_realm, num_total);
+            test_realm->sync_session()->pause(); // Perform the local updates offline
```

Comment:
From my experience, this can cause a race. You're better using `shutdown_and_wait()`.